File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/JdbcInsertFlushFunction.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.destination.jdbc;

import io.airbyte.cdk.integrations.destination.async.function.DestinationFlushFunction;
import io.airbyte.cdk.integrations.destination.async.partial_messages.PartialAirbyteMessage;
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.RecordWriter;
import io.airbyte.cdk.integrations.destination.jdbc.constants.GlobalDataSizeConstants;
import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
import io.airbyte.protocol.models.v0.StreamDescriptor;
import java.util.stream.Stream;

public class JdbcInsertFlushFunction implements DestinationFlushFunction {

  private final RecordWriter<PartialAirbyteMessage> recordWriter;

  public JdbcInsertFlushFunction(final RecordWriter<PartialAirbyteMessage> recordWriter) {
    this.recordWriter = recordWriter;
  }

  @Override
  public void flush(final StreamDescriptor desc, final Stream<PartialAirbyteMessage> stream) throws Exception {
    recordWriter.accept(
        new AirbyteStreamNameNamespacePair(desc.getName(), desc.getNamespace()),
        stream.toList());
  }

  @Override
  public long getOptimalBatchSizeBytes() {
    // TODO tune this value - currently SqlOperationUtils partitions 10K records per insert statement,
    // but we'd like to stop doing that and instead control sql insert statement size via batch size.
    return GlobalDataSizeConstants.DEFAULT_MAX_BATCH_SIZE_BYTES;
  }

}

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/JdbcInsertFlushFunction.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.jdbc

import io.airbyte.cdk.integrations.destination.async.function.DestinationFlushFunction
import io.airbyte.cdk.integrations.destination.async.model.PartialAirbyteMessage
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.RecordWriter
import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair
import io.airbyte.protocol.models.v0.StreamDescriptor
import java.util.stream.Stream

class JdbcInsertFlushFunction(
    private val recordWriter: RecordWriter<PartialAirbyteMessage>,
    override val optimalBatchSizeBytes: Long
) : DestinationFlushFunction {
    @Throws(Exception::class)
    override fun flush(streamDescriptor: StreamDescriptor, stream: Stream<PartialAirbyteMessage>) {
        recordWriter.accept(
            AirbyteStreamNameNamespacePair(streamDescriptor.name, streamDescriptor.namespace),
            stream.toList()
        )
    }
}
